feat: Interfaces and docgen (SC-2343)

deluca-mike opened the pull request:
# Description

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog
## Function Signature Changes

## Features 

## Events

deluca-mike committed:
refactor: Interfaces, Natspec, flatten script

deluca-mike committed:
chore: Update Metadata.json

satyamakgec commented on contracts/core/chainlink-oracle/v1/ChainlinkOracle.sol:
```diff
@@ -7,39 +7,33 @@ import "external-interfaces/IChainlinkAggregatorV3.sol";
 
-    bool   public manualOverride;
-    int256 public manualPrice;
+    address public override immutable assetAddress;
```
Why do we need `override` here? My assumption here is that it is needed to avoid shadow declaration of the same variable name.

lucas-manuel replied:
If we remove `override` we get this type of error:
```
contracts/core/loan/v1/Loan.sol:35:5: TypeError: Overriding public state variable is missing "override" specifier.
    State public loanState;
    ^--------------------^
contracts/core/loan/v1/interfaces/ILoan.sol:90:5: Overridden public state variable is here:
    function loanState() external view returns (State);
    ```

deluca-mike replied:
Needs to override the interface (function) defined in `IChainlinkOracle.sol`.

satyamakgec commented on contracts/core/chainlink-oracle/v1/ChainlinkOracle.sol:
```diff
@@ -7,39 +7,33 @@ import "external-interfaces/IChainlinkAggregatorV3.sol";
-        @return price The latest price.
-    */
-    function getLatestPrice() public view returns (int256) {
+    function getLatestPrice() public override view returns (int256) {
```
Can't we have some CI job that doesn't the Interface verification stuff instead of adding here. It is a pain to keep `override` without any use of it.

lucas-manuel replied:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/4d0f8c1da8654a478f046ea7cf83d2166e1025af/contracts/token/ERC20/ERC20.sol#L111

I actually like the `override` keyword, it locks us into the interface function definition.

lucas-manuel replied:
If you define a function with `override` and it does not have a corresponding base function, it will error:
```
contracts/core/loan/v1/Loan.sol:208:38: TypeError: Function has override specified but does not override anything.
    function testFunction() external override {}
```

lucas-manuel replied:
This error also comes up if the function signature changes:
```
contracts/core/loan/v1/Loan.sol:200:53: TypeError: Function has override specified but does not override anything.
    function makeFullPayment(address test) external override {
```

deluca-mike replied:
Agreed it sucks. IMO its a fault of Solidity that one needs to override interfaces at all. The fact that interface prototypes are implicitly `virtual` should mean that implementation functions implicitly override those inherited interface definitions. However, putting `override` here and there, as we write code, is not that bad, and I don't think it should be a CI job. We do manually formatting, might as well type a handful of words manually.

satyamakgec commented on contracts/core/chainlink-oracle/v1/ChainlinkOracle.sol:
```diff
@@ -49,53 +43,30 @@ contract ChainlinkOracle is Ownable {
-        require(aggregator != address(0), "CO:ZERO_AGGREGATOR_ADDR");
-        emit ChangeAggregatorFeed(aggregator, address(priceFeed));
-        priceFeed = IChainlinkAggregatorV3(aggregator);
+    function changeAggregator(address newAggregator) external override onlyOwner {
```
I hate to break it but it will reduce the code readability a lot if we keep every doc in the interfaces.

lucas-manuel replied:
I think that if we make a note of this in the README it will be fine. I prefer moving the comments out into the interface, makes the code neater and smaller, and if a dev needs to refer to the NatSpec they can just search the function name and it'll show up.

deluca-mike replied:
I have/had similar feelings, but on the flipside, those who only import our interfaces will see the docs. If we put the docs in the implementations only, then interfaces will lack any info for importers, who probably need it more (they don't know the system as well as us). What we could do is some CI task or script to copy docs from each to their couterpart.

satyamakgec commented on metadata.json:
```diff
@@ -2,87 +2,87 @@
   "contractName": "BasicFDT",
   "contractSize": 0,
-  "sourceHash": "0x67502ea16139de7c6a7d24101169d931973253f76a2a85bbc4a63da4fa8e6123",
+  "sourceHash": "0xd718410027ac8b1e9f557334d8236cca88ebe89c028eaddced0c42d03e608be4",
```
Is it essential to push this file?

deluca-mike replied:
No, it is not. Ideally CI will build the old and new and allow them to be compared as a requirement for merging a PR.

lucas-manuel commented on contracts/core/chainlink-oracle/v1/interfaces/IChainlinkOracle.sol:
```diff
@@ -0,0 +1,96 @@
+import "core/globals/v1/interfaces/IMapleGlobals.sol";
+
+/// @title ChainlinkOracle is a wrapper contract for Chainlink oracle price feeds that allows for manual price feed overrides.
+interface IChainlinkOracle {
```
I love this 👍

lucas-manuel commented on contracts/core/collateral-locker/v1/CollateralLockerFactory.sol:
```diff
@@ -1,25 +1,19 @@
+
 import "./CollateralLocker.sol";
 
 /// @title CollateralLockerFactory instantiates CollateralLockers.
```
Does it matter that we have two titles?

deluca-mike replied:
No.

lucas-manuel commented on contracts/core/funding-locker/v1/FundingLocker.sol:
```diff
@@ -8,8 +8,8 @@ contract FundingLocker {
 
-    IERC20  public immutable liquidityAsset;  // Asset the Loan was funded with.
-    address public immutable loan;            // Loan this FundingLocker has funded.
+    IERC20  public immutable liquidityAsset;
+    address public immutable loan;
```
This needs to inherit IFundingLocker and have overrides

lucas-manuel commented on contracts/core/funding-locker/v1/FundingLockerFactory.sol:
```diff
@@ -6,19 +6,13 @@ import "./FundingLocker.sol";
 contract FundingLockerFactory {
 
-    mapping(address => address) public owner;     // Mapping of FundingLocker addresses to their owner (i.e owner[locker] = Owner of the FundingLocker).
+    mapping(address => address) public owner;     // Owners of respective FundingLockers.
```
Needs to inherit IFundingLockerFactory and have overrides

lucas-manuel commented on contracts/core/funds-distribution-token/v1/BasicFDT.sol:
```diff
@@ -40,7 +37,7 @@ abstract contract BasicFDT is IBaseFDT, ERC20 {
                    and try to distribute it in the next distribution.
-    */
+     */
     function _distributeFunds(uint256 value) internal {
```
Does the NatSpec for internal functions get generated in the docgen?

deluca-mike replied:
No.

lucas-manuel commented on contracts/core/funds-distribution-token/v1/interfaces/IBasicFDT.sol:
```diff
@@ -5,16 +5,44 @@ import "lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+        @param account          The address of some account.
+        @param pointsCorrection The new value of the account's `pointsCorrection`.
+     */
+    event PointsCorrectionUpdated(address indexed account, int256 pointsCorrection);
```
@deluca-mike did you change any names of event params in the definitions? I saw that some of the function param names changed

lucas-manuel replied:
If so, we will have to communicate all changes to the offchain team

deluca-mike replied:
Will make sure none change.

lucas-manuel replied:
Current is: `event PointsCorrectionUpdated(address indexed, int256);`

lucas-manuel commented on contracts/core/funds-distribution-token/v1/interfaces/IBasicFDT.sol:
```diff
@@ -5,16 +5,44 @@ import "lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
 import "./IBaseFDT.sol";
 
+/// @title BasicFDT implements the basic level FDT functionality for accounting for revenues.
 interface IBasicFDT is IBaseFDT, IERC20 {
```
Should we consolidate IBaseFDT and IBasicFDT?

deluca-mike replied:
Will try.

lucas-manuel reviewed (commented):
Partial review, will continue after questions are answered

deluca-mike committed:
chore: PR Review

deluca-mike committed:
fix: Oracle Interface

lucas-manuel commented on contracts/core/collateral-locker/v1/interfaces/ICollateralLocker.sol:
```diff
@@ -1,12 +1,27 @@
 
-    function pull(address, uint256) external;
+    /**
+        @dev   Transfers `amount` of Collateral Asset to `destination`. 
```
`amt` `dst`

lucas-manuel commented on contracts/core/chainlink-oracle/v1/interfaces/IChainlinkOracle.sol:
```diff
@@ -0,0 +1,97 @@
+        @param oldPrice The old price.
+        @param newPrice The new price.
+     */
+    event SetManualPrice(int256 oldPrice, int256 newPrice);
```
`_oldPrice` `_newPrice`

lucas-manuel commented on contracts/core/chainlink-oracle/v1/interfaces/IChainlinkOracle.sol:
```diff
@@ -0,0 +1,97 @@
+interface IChainlinkOracle is IOracle {
+
+    /**
+        @dev   Emits an event indicating that the price feed aggregator has changed from `oldFeed` to `newFeed`.
```
`_oldMedianizer` `_newMedianizer`

lucas-manuel commented on contracts/core/funding-locker/v1/interfaces/IFundingLocker.sol:
```diff
@@ -1,14 +1,33 @@
 
-    function pull(address, uint256) external;
+    /**
+        @dev   Transfers `amount` of Liquidity Asset to `destination`. 
```
`amt` `destination`

lucas-manuel commented on contracts/core/funds-distribution-token/v1/interfaces/IBasicFDT.sol:
```diff
@@ -3,18 +3,71 @@ pragma solidity 0.6.11;
+        @dev   This event emits when the internal `pointsPerShare` is updated.
+        @param pointsPerShare The new value of the internal `pointsPerShare`.
+     */
+    event PointsPerShareUpdated(uint256 pointsPerShare);
```
Current `pointsPerShareUpdated` is `event PointsPerShareUpdated(uint256)`

lucas-manuel commented on contracts/core/funds-distribution-token/v1/interfaces/IExtendedFDT.sol:
```diff
@@ -3,24 +3,66 @@ pragma solidity 0.6.11;
+        @dev   This event emits when the internal `lossesPerShare` is updated.
+        @param lossesPerShare The new value of the internal `lossesPerShare`.
+     */
+    event LossesPerShareUpdated(uint256 lossesPerShare);
```
Current: `event LossesPerShareUpdated(uint256);`

lucas-manuel commented on contracts/core/funds-distribution-token/v1/interfaces/IExtendedFDT.sol:
```diff
@@ -3,24 +3,66 @@ pragma solidity 0.6.11;
+        @param lossesRecognized      The amount of losses that were recognized.
+        @param totalLossesRecognized The total amount of losses that are recognized.
+     */
+    event LossesRecognized(address indexed by, uint256 lossesRecognized, uint256 totalLossesRecognized);
```
Current: 
```js
event LossesCorrectionUpdated(address indexed, int256);
event LossesDistributed(address indexed, uint256);
event LossesRecognized(address indexed, uint256, uint256);
```

lucas-manuel commented on contracts/core/oracle/v1/Oracle.sol:
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.6.11;
+
+import "./interfaces/IOracle.sol";
+
+/// @title Oracle is a price oracle feed.
+abstract contract Oracle is IOracle {
+
+    function getLatestPrice() external override virtual view returns (int256);
+
+}
```
What are we using this file for?

deluca-mike replied:
You're right. Getting rid of it.

lucas-manuel commented on contracts/core/treasury/v1/interfaces/IMapleTreasury.sol:
```diff
@@ -1,22 +1,69 @@
+/// @title MapleTreasury earns revenue from Loans and distributes it to token holders and the Maple development team.
 interface IMapleTreasury {
 
+    event ERC20Conversion(address indexed asset, uint256 amountIn, uint256 amountOut);
+
+    event DistributedToHolders(uint256 amount);
+    
+    event ERC20Reclaimed(address indexed asset, uint256 amount);
+    
+    event GlobalsSet(address newGlobals);
```
Events need NatSpec

lucas-manuel commented on metadata.json:
```diff
@@ -2,215 +2,207 @@
-  "bytecodeHashWithoutLibRefs": "0x2ec1261d0d7cfa0d01decfde857a678a9f84c409068a76b893143a934c4e5901",
-  "rawBytecode": "608060405234801561001057600080fd5b506004361061004c5760003560e01c806319eb783a146100515780632ec63d7c1461009357806364e1fd55146100cd578063666e1b39146100eb575b600080fd5b6100776004803603602081101561006757600080fd5b50356001600160a01b0316610111565b604080516001600160a01b039092168252519081900360200190f35b6100b9600480360360208110156100a957600080fd5b50356001600160a01b03166101d9565b604080519 ...
+  "sourceHash": "0x415a7ce1341839a7f3f09b4f3e53dbaf1ae01e9d0558ffee001b03e273882d5f",
+  "bytecodeHashWithLibRefs": "0xbe67b0846442c3f5603c5541a7a3f5b3caae3f84761490baa4b5e387ab8cfb1c",
+  "bytecodeHashWithoutLibRefs": "0xbe67b0846442c3f5603c5541a7a3f5b3caae3f84761490baa4b5e387ab8cfb1c",
```
Why did this change?

deluca-mike replied:
Factory contains two IPFS hashes.

lucas-manuel commented on metadata.json:
```diff
@@ -2,215 +2,207 @@
-  "bytecodeHashWithoutLibRefs": "0xb1340e449097a838168b4d677bb30a7dbee3452f3493a054ae07ffade39f88ce",
-  "rawBytecode": "608060405234801561001057600080fd5b506004361061004c5760003560e01c806319eb783a146100515780632ec63d7c1461009357806364e1fd55146100cd578063666e1b39146100eb575b600080fd5b6100776004803603602081101561006757600080fd5b50356001600160a01b0316610111565b604080516001600160a01b039092168252519081900360200190f35b6100b9600480360360208110156100a957600080fd5b50356001600160a01b03166101d9565b604080519 ...
+  "sourceHash": "0x005ae97e29bd14b8a7774b965c12e7b1dbe3911271259dd7ee4142a7c18c798c",
+  "bytecodeHashWithLibRefs": "0x7d5f83d24bd3113e687ca7b051408ce13d2e17d70d99c30ad999e831ae0d7e26",
+  "bytecodeHashWithoutLibRefs": "0x7d5f83d24bd3113e687ca7b051408ce13d2e17d70d99c30ad999e831ae0d7e26",
```
Why did this change?

deluca-mike replied:
Factory contains two IPFS hashes.

lucas-manuel commented on metadata.json:
```diff
@@ -2,215 +2,207 @@
-  "bytecodeHashWithoutLibRefs": "0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
-  "rawBytecode": ""
+  "sourceHash": "0x3359045f58403ee82a40c784c6f8c5615346e9025b578ac2d646739a3a67fb0c",
+  "bytecodeHashWithLibRefs": "0x25617a447ef02af53d6adce7d9ea020090fa2ab8fcc9031133d403a450e052f8",
+  "bytecodeHashWithoutLibRefs": "0x25617a447ef02af53d6adce7d9ea020090fa2ab8fcc9031133d403a450e052f8",
```
Why did this change?

deluca-mike replied:
Factory contains two IPFS hashes.

lucas-manuel commented on metadata.json:
```diff
@@ -328,195 +320,203 @@
-  "bytecodeHashWithoutLibRefs": "0x8f5925ab08dcf0e6b3b2d373256762ac326ec0a9911497aa4e408edcd5c415e2",
-  "rawBytecode": "608060405234801561001057600080fd5b506004361061004c5760003560e01c806319eb783a146100515780632ec63d7c1461009357806364e1fd55146100cd578063666e1b39146100eb575b600080fd5b6100776004803603602081101561006757600080fd5b50356001600160a01b0316610111565b604080516001600160a01b039092168252519081900360200190f35b6100b9600480360360208110156100a957600080fd5b50356001600160a01b03166101d9565b604080519 ...
+  "sourceHash": "0xf7a163a9b34b485130ae15f2d0b2939e5dfb66c55bf1ee4edca49eaa1324498e",
+  "bytecodeHashWithLibRefs": "0x4c1bb131e702750a1ddebf99c82b7b6d4169c8d2ae72ab923e699ec1838cdc11",
+  "bytecodeHashWithoutLibRefs": "0x4c1bb131e702750a1ddebf99c82b7b6d4169c8d2ae72ab923e699ec1838cdc11",
```
Why did this change?

deluca-mike replied:
Factory contains two IPFS hashes.

lucas-manuel commented on metadata.json:
```diff
@@ -328,195 +320,203 @@
-  "bytecodeHashWithoutLibRefs": "0xced5a090306465e3cdb07b71e2afea5a610214bf5d3897ad99f63cd47f9e4696",
-  "rawBytecode": "60806040523480156200001157600080fd5b5060043610620001185760003560e01c80638c38922f11620000a5578063d5ab2074116200006f578063d5ab20741462000261578063e1ec3c681462000309578063e70dd6cf1462000329578063ee9b75b514620003335762000118565b80638c38922f14620001df578063a620e0ac14620001e9578063c31245251462000212578063cc2e0a2614620002385762000118565b80633f4ba83a11620000e75780633f4ba83a14620001b7578 ...
+  "sourceHash": "0xd5753d5c7a0ecf9479eb26ac0e010d3143e9e975e87a655596f743a8a7f33d04",
+  "bytecodeHashWithLibRefs": "0x5b046aa1589581ae8366b01dae7f2d89215481ca9184dcc4cc75d2ddbd14120f",
+  "bytecodeHashWithoutLibRefs": "0x5b046aa1589581ae8366b01dae7f2d89215481ca9184dcc4cc75d2ddbd14120f",
```
Why did this change?

deluca-mike replied:
Factory contains two IPFS hashes.

lucas-manuel commented on metadata.json:
```diff
@@ -328,195 +320,203 @@
+  "sourceHash": "0xf88fe1f83903666b819a246f974a7b561fdc2664627dd88a137f20305e8620bd",
   "bytecodeHashWithLibRefs": "0x9ce099a9c37022a12e5465ef1cb51b37af9b4e9a332d25f3c1b985ea70863038",
   "bytecodeHashWithoutLibRefs": "0x9ce099a9c37022a12e5465ef1cb51b37af9b4e9a332d25f3c1b985ea70863038",
-  "rawBytecode": "608060405234801561001057600080fd5b50600436106101ce5760003560e01c80637b0a47ee11610104578063c8f33c91116100a2578063df136d6511610071578063df136d65146103fb578063e9fad8ee14610403578063ebe2b12b1461040b578063f2fde38b14610413576101ce565b8063c8f33c91146103c6578063cc1a378f146103ce578063cd3daf9d146103eb578063d1af0c7d146103f3576101ce565b80638b876347116100de5780638b876347146103735780638da5cb5 ...
+  "rawBytecode": "608060405234801561001057600080fd5b50600436106101ce5760003560e01c80637b0a47ee11610104578063c8f33c91116100a2578063df136d6511610071578063df136d65146103fb578063e9fad8ee14610403578063ebe2b12b1461040b578063f2fde38b14610413576101ce565b8063c8f33c91146103c6578063cc1a378f146103ce578063cd3daf9d146103eb578063d1af0c7d146103f3576101ce565b80638b876347116100de5780638b876347146103735780638da5cb5 ...
```
Why did this change?

lucas-manuel replied:
Why did this change?

lucas-manuel replied:
Why did this change?

lucas-manuel commented on flatten.sh:
```diff
@@ -1,33 +1,33 @@
-    echo "Flattening: $filename"
-    hevm flatten --source-file $filepath | sed '/SPDX-License-Identifier/d' | sed -e '1s/^/\/\/ SPDX-License-Identifier:  AGPL-3.0-or-later /' | sed -r 'N;s/(\/\/\/\/\/\/ .*)\n\s*$/\1/;P;D' > "flattened-contracts/$filename"
-    cat "flattened-contracts/$filename" | grep -E ") external |) public" > "flattened-contracts/functions/$filename"
+    mkdir -p "flattened-contracts/${filename%/*}"
```
Why are we making new directories for each iteration in the for loop?

deluca-mike replied:
hevm flattener requires the output directory to exist, as it will only create directories within it.

If you remove that line, you get:
`./flatten.sh: line 14: flattened-contracts/core/chainlink-oracle/v1/ChainlinkOracle.sol: No such file or directory`

lucas-manuel reviewed (changes requested):
Some comments/questions/changes

deluca-mike committed:
refactor: PR Review 2

- reverted back to some poorly named parameters
- reverted the creation/use of an abstract Oracle
- created standalone ISubFactory for inheritance
- corrected some natspec
- added complete natspec to StakeLocker and StakeLockerfactory
- moved IMapleTokon to external-interfaces
- removed 3 interfaces from being built on thier own

Commit:
feat: update cache commit
